libcontainer/capabilities/capabilities.go

```diff
@@ -0,0 +1,158 @@
+//go:build linux
+
+package capabilities
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"syscall"
+
+	"github.com/moby/sys/capability"
+	"github.com/opencontainers/runc/libcontainer/configs"
+	"github.com/sirupsen/logrus"
+)
+
+func capToStr(c capability.Cap) string {
+	return "CAP_" + strings.ToUpper(c.String())
+}
+
+var capMap = sync.OnceValues(func() (map[string]capability.Cap, error) {
+	list, err := capability.ListSupported()
+	if err != nil {
+		return nil, err
+	}
+	cm := make(map[string]capability.Cap, len(list))
+	for _, c := range list {
+		cm[capToStr(c)] = c
+	}
+	return cm, nil
+})
+
+// KnownCapabilities returns the list of the known capabilities.
+// Used by `runc features`.
+func KnownCapabilities() []string {
+	list := capability.ListKnown()
+	res := make([]string, len(list))
+	for i, c := range list {
+		res[i] = "CAP_" + strings.ToUpper(c.String())
+	}
+	return res
+}
+
+// New creates a new Caps from the given Capabilities config. Unknown Capabilities
+// or Capabilities that are unavailable in the current environment are ignored,
+// printing a warning instead.
+func New(capConfig *configs.Capabilities) (*Caps, error) {
+	var c Caps
+	if capConfig == nil {
+		return &c, nil
+	}
+
+	_, err := capMap()
+	if err != nil {
+		return nil, err
+	}
+	unknownCaps := make(map[string]struct{})
+	c.caps = map[capability.CapType][]capability.Cap{
+		capability.BOUNDING:    capSlice(capConfig.Bounding, unknownCaps),
+		capability.EFFECTIVE:   capSlice(capConfig.Effective, unknownCaps),
+		capability.INHERITABLE: capSlice(capConfig.Inheritable, unknownCaps),
+		capability.PERMITTED:   capSlice(capConfig.Permitted, unknownCaps),
+		capability.AMBIENT:     capSlice(capConfig.Ambient, unknownCaps),
+	}
+	if c.pid, err = capability.NewPid2(0); err != nil {
+		return nil, err
+	}
+	if len(unknownCaps) > 0 {
+		logrus.Warn("ignoring unknown or unavailable capabilities: ", mapKeys(unknownCaps))
```

Comment:
Any particular reason for not using `slices.Sorted(maps.Keys(unknownCaps))` (and therefore avoid the `mapKeys()` function)?

Reply:
no not really, slices.Sorted seems more efficient, will fix